scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,266 @@
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+    upper_timestamp: Option<chrono::Duration>,
+    sleep_interval: time::Duration,
+}
+
+impl Reader {
+    pub fn new(
+        stream_ids: Vec<StreamID>,
+        start_timestamp: chrono::Duration,
+        window_size: chrono::Duration,
+        safety_interval: chrono::Duration,
+        sleep_interval: time::Duration,
+    ) -> Reader {
+        Reader {
+            stream_id_vec: stream_ids,
+            lower_timestamp: start_timestamp,
+            window_size,
+            safety_interval,
+            upper_timestamp: std::option::Option::None,
+            sleep_interval,
+        }
+    }
+
+    // Mutable access.
+    fn lower_timestamp_mut(&mut self) -> Option<chrono::Duration> {
+        self.upper_timestamp
+    }
+
+    pub fn set_lower_timestamp(&mut self, new_lower_timestamp: chrono::Duration) {
+        self.upper_timestamp = Some(new_lower_timestamp);
+    }
+
+    pub async fn fetch_cdc(
+        &self,
+        keyspace: String,
+        table_name: &str,
+    ) -> Result<(), Box<dyn Error>> {
+        const URI: &str = "172.17.0.2:9042";
+        let session: Session = SessionBuilder::new().known_node(URI).build().await.unwrap(); // Will panic in case of error.
+        let query = format!(
+            "SELECT * FROM {}.{}_scylla_cdc_log \
+            WHERE \"cdc$stream_id\" in (?) \
+            AND \"cdc$time\" >= maxTimeuuid(?) \
+            AND \"cdc$time\" < minTimeuuid(?)",
+            keyspace, table_name
+        );
+        let query_base = session.prepare(query).await.unwrap();
+        let mut continue_fetch = true;
+        let mut lower_timestamp = self.lower_timestamp;
+
+        while continue_fetch {
+            let now = chrono::Duration::from_std(
+                SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .unwrap(),
+            )
+            .unwrap();
+            let upper_timestamp;
+
+            match self.upper_timestamp {
+                Some(upper) => upper_timestamp = upper,
+                None => {
+                    upper_timestamp = max(
+                        lower_timestamp,
+                        min(
+                            lower_timestamp + self.window_size,
+                            now - self.safety_interval,
+                        ),
+                    )
+                }
+            }
+
+            if lower_timestamp == upper_timestamp {
+                continue_fetch = false;
+            }
+
+            let mut rows_stream = session
+                .execute_iter(
+                    query_base.clone(),
+                    (
+                        &self.stream_id_vec,
+                        Timestamp(lower_timestamp),
+                        Timestamp(upper_timestamp),
+                    ),
+                )
+                .await?
+                .into_typed::<(i32, i32, String, String)>();
```

Comment:
This cast is incorrect - there are much more columns selected in your query.

Reply:
You shouldn't perform any casts here. The stream will return `Row`s which are untyped wrappers around `Vec<CqlValue>` - please pass the rows directly to the callback (with which you should replace printing the rows on line 124).